File: src/main/java/io/github/saneea/dvh/feature/binary/gzip/FromGzip.java
package io.github.saneea.dvh.feature.binary.gzip;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.zip.GZIPInputStream;

import io.github.saneea.dvh.Feature;
import io.github.saneea.dvh.FeatureContext;

public class FromGzip implements//
		Feature, //
		Feature.In.Bin.Stream, //
		Feature.Out.Bin.Stream {

	private InputStream in;
	private OutputStream out;

	@Override
	public Meta meta(FeatureContext context) {
		return Meta.from("extract from GZIP");
	}

	@Override
	public void run(FeatureContext context) throws IOException {
		try (GZIPInputStream gzipIn = new GZIPInputStream(in)) {
			gzipIn.transferTo(out);
		}
	}

	@Override
	public void setIn(InputStream in) {
		this.in = in;
	}

	@Override
	public void setOut(OutputStream out) {
		this.out = out;
	}
}

File: src/main/java/io/github/saneea/dvh/feature/binary/gzip/FromGzip.kt
package io.github.saneea.dvh.feature.binary.gzip

import io.github.saneea.dvh.Feature
import io.github.saneea.dvh.Feature.Meta
import io.github.saneea.dvh.FeatureContext
import io.github.saneea.dvh.utils.transferFrom
import java.io.InputStream
import java.io.OutputStream
import java.util.zip.GZIPInputStream

class FromGzip :
    Feature,
    Feature.In.Bin.Stream,
    Feature.Out.Bin.Stream {

    override lateinit var context: FeatureContext
    override lateinit var inBinStream: InputStream
    override lateinit var outBinStream: OutputStream

    override val meta = Meta("extract from GZIP")

    override fun run() {
        GZIPInputStream(inBinStream).use(outBinStream::transferFrom)
    }
}